File: innerlibs1/FluentSwitch.vb

Imports System.Runtime.CompilerServices

Public Module FluentSwitchExt

    <Extension()>
    Public Function Switch(Of TestType, ReturnType)(Input As TestType) As FluentSwitch(Of TestType, ReturnType)
        Return New FluentSwitch(Of TestType, ReturnType)(Input)
    End Function

    <Extension()>
    Public Function Switch(Of TestType, ReturnType)(Input As TestType, Test As Action(Of FluentSwitch(Of TestType, ReturnType))) As ReturnType
        Dim a = Input.Switch(Of ReturnType)()
        Test(a)
        Return a.GetValue()
    End Function
End Module

Public Class FluentSwitch(Of TestType, ReturnType)

    Private dic As New Dictionary(Of TestType, ReturnType)

    Private input As TestType = Nothing
    Private defaultv As ReturnType = Nothing

    Public Sub New(Input As TestType, Optional DefaultValue As ReturnType = Nothing)
        Me.input = Input
        Me.defaultv = DefaultValue
    End Sub

    Public Function [Case](Value As TestType, ReturnValue As ReturnType) As FluentSwitch(Of TestType, ReturnType)
        Return [Case]({Value}, ReturnValue)
    End Function
    Public Function [Case](Values As IEnumerable(Of TestType), ReturnValue As ReturnType) As FluentSwitch(Of TestType, ReturnType)
        For Each item As TestType In Values
            dic.Set(item, ReturnValue)
        Next
        Return Me
    End Function


    Public Function [Default](ReturnValue As ReturnType) As FluentSwitch(Of TestType, ReturnType)
        defaultv = ReturnValue
        Return Me
    End Function

    Public Function GetValue() As ReturnType
        Return dic.GetValueOr(input, defaultv)
    End Function

    Public Shared Widening Operator CType(FS As FluentSwitch(Of TestType, ReturnType)) As ReturnType
        If FS Is Nothing Then Return Nothing
        Return FS.GetValue()
    End Operator

End Class

File: InnerLibs/FluentSwitch.cs
using System;
using System.Collections.Generic;

namespace InnerLibs
{
    public static class FluentSwitchExt
    {
        public static FluentSwitch<T1, T2> Switch<T2, T1>(this T1 Input)
        {
            return new FluentSwitch<T1, T2>(Input);
        }

        public static T2 Switch<T2, T1>(this T1 Input, Action<FluentSwitch<T1, T2>> Test)
        {
            var a = Input.Switch<T2, T1>();
            Test(a);
            return a.GetValue();
        }
    }

    public class FluentSwitch<T1, T2>
    {
        private Dictionary<T1, T2> dic = new Dictionary<T1, T2>();
        private T1 input = default;
        private T2 defaultv = default;

        public FluentSwitch(T1 Input, T2 DefaultValue = default)
        {
            input = Input;
            defaultv = DefaultValue;
        }

        public FluentSwitch<T1, T2> Case(T1 Value, T2 ReturnValue)
        {
            return Case(new[] { Value }, ReturnValue);
        }

        public FluentSwitch<T1, T2> Case(IEnumerable<T1> Values, T2 ReturnValue)
        {
            foreach (T1 item in Values)
                dic.Set(item, ReturnValue);
            return this;
        }

        public FluentSwitch<T1, T2> Default(T2 ReturnValue)
        {
            defaultv = ReturnValue;
            return this;
        }

        public T2 GetValue()
        {
            return dic.GetValueOr(input, defaultv);
        }

        public static implicit operator T2(FluentSwitch<T1, T2> FS)
        {
            if (FS is null)
                return default;
            return FS.GetValue();
        }
    }
}